### PR TITLE
PR 13757 follow-up: add missing with-discountOverlaps Similarity constructor variants, CHANGES.txt entries

### PR DESCRIPTION
**Pull request by cpoerschke**
#13757 follow-up

### CODE REVIEW

**Comment by cpoerschke** [comment]
Thanks @psalagnac for identifying this in the https://issues.apache.org/jira/browse/SOLR-17471 ticket!

File: lucene/core/src/java/org/apache/lucene/search/similarities/Axiomatic.java [hunk 1]
```diff
@@ -44,13 +44,26 @@ public abstract class Axiomatic extends SimilarityBase {
...
+
+  /**
+   * Constructor setting all Axiomatic hyperparameters
+   *
+   * @param discountOverlaps true if overlap tokens should not impact document length for scoring.
+   * @param s hyperparam for the growth function
+   * @param queryLen the query length
+   * @param k hyperparam for the primitive weighting function
+   */
+  public Axiomatic(boolean discountOverlaps, float s, int queryLen, float k) {
```
Comment (cpoerschke) [inline lucene/core/src/java/org/apache/lucene/search/similarities/Axiomatic.java:65]: Technically additional variant constructors could also be added for all the classes that derive from `Axiomatic` but there are relatively many derivations with multiple constructors. The class is marked `@lucene.experimental` and so perhaps it's okay to add additional variants if/when a need for them is identified?

File: lucene/core/src/java/org/apache/lucene/search/similarities/DFISimilarity.java [hunk 2]
```diff
@@ -46,11 +46,23 @@ public class DFISimilarity extends SimilarityBase {
...
+    this(independenceMeasure, true);
+  }
+
+  /**
+   * Create DFI with the specified parameters
+   *
+   * @param independenceMeasure measure of divergence from independence
+   * @param discountOverlaps true if overlap tokens should not impact document length for scoring.
+   */
+  public DFISimilarity(Independence independenceMeasure, boolean discountOverlaps) {
```
Comment (cpoerschke) [inline lucene/core/src/java/org/apache/lucene/search/similarities/DFISimilarity.java:64]: to support replacement of this: https://github.com/apache/solr/blob/releases/solr/9.7.0/solr/core/src/java/org/apache/solr/search/similarities/DFISimilarityFactory.java#L62-L63

File: lucene/core/src/java/org/apache/lucene/search/similarities/DFRSimilarity.java [hunk 3]
```diff
@@ -98,7 +98,7 @@ public DFRSimilarity(
   }
 
   /**
-   * Creates DFRSimilarity from the three components.
+   * Creates DFRSimilarity from the four components.
```
Comment (cpoerschke) [inline lucene/core/src/java/org/apache/lucene/search/similarities/DFRSimilarity.java:101]: to support replacement of this: https://github.com/apache/solr/blob/releases/solr/9.7.0/solr/core/src/java/org/apache/solr/search/similarities/DFRSimilarityFactory.java#L163-L164

File: lucene/core/src/java/org/apache/lucene/search/similarities/IBSimilarity.java [hunk 4]
```diff
@@ -86,6 +86,26 @@ public class IBSimilarity extends SimilarityBase {
...
+   * @param distribution probabilistic distribution modeling term occurrence
+   * @param lambda distribution's &lambda;<sub>w</sub> parameter
+   * @param normalization term frequency normalization
+   * @param discountOverlaps true if overlap tokens should not impact document length for scoring.
+   */
+  public IBSimilarity(
+      Distribution distribution,
+      Lambda lambda,
+      Normalization normalization,
+      boolean discountOverlaps) {
```
Comment (cpoerschke) [inline lucene/core/src/java/org/apache/lucene/search/similarities/IBSimilarity.java:107]: to support replacement of this: https://github.com/apache/solr/blob/releases/solr/9.7.0/solr/core/src/java/org/apache/solr/search/similarities/IBSimilarityFactory.java#L103-L104

File: lucene/core/src/java/org/apache/lucene/search/similarities/LMDirichletSimilarity.java [hunk 5]
```diff
@@ -49,6 +55,12 @@ public LMDirichletSimilarity(CollectionModel collectionModel, float mu) {
 
   /** Instantiates the similarity with the provided &mu; parameter. */
   public LMDirichletSimilarity(float mu) {
+    this(true, mu);
+  }
+
+  /** Instantiates the similarity with the provided parameters. */
+  public LMDirichletSimilarity(boolean discountOverlaps, float mu) {
```
Comment (cpoerschke) [inline lucene/core/src/java/org/apache/lucene/search/similarities/LMDirichletSimilarity.java:62]: to support replacement of this: https://github.com/apache/solr/blob/releases/solr/9.7.0/solr/core/src/java/org/apache/solr/search/similarities/LMDirichletSimilarityFactory.java#L54-L56

File: lucene/core/src/java/org/apache/lucene/search/similarities/LMDirichletSimilarity.java [hunk 6]
```diff
@@ -62,8 +74,15 @@ public LMDirichletSimilarity(CollectionModel collectionModel) {
   }
 
   /** Instantiates the similarity with the default &mu; value of 2000. */
+  public LMDirichletSimilarity(boolean discountOverlaps) {
```
Comment (cpoerschke) [inline lucene/core/src/java/org/apache/lucene/search/similarities/LMDirichletSimilarity.java:77]: to support replacement of this: https://github.com/apache/solr/blob/releases/solr/9.7.0/solr/core/src/java/org/apache/solr/search/similarities/LMDirichletSimilarityFactory.java#L54-L56

File: lucene/core/src/java/org/apache/lucene/search/similarities/LMJelinekMercerSimilarity.java [hunk 7]
```diff
@@ -52,6 +58,12 @@ public LMJelinekMercerSimilarity(CollectionModel collectionModel, float lambda)
 
   /** Instantiates with the specified &lambda; parameter. */
   public LMJelinekMercerSimilarity(float lambda) {
+    this(true, lambda);
+  }
+
+  /** Instantiates with the specified parameters. */
+  public LMJelinekMercerSimilarity(boolean discountOverlaps, float lambda) {
```
Comment (cpoerschke) [inline lucene/core/src/java/org/apache/lucene/search/similarities/LMJelinekMercerSimilarity.java:65]: to support replacement of this: https://github.com/apache/solr/blob/releases/solr/9.7.0/solr/core/src/java/org/apache/solr/search/similarities/LMJelinekMercerSimilarityFactory.java#L55-L56

**Comment by cpoerschke** [comment]
w.r.t. the CI failure -- https://github.com/apache/lucene/actions/runs/11130252581/job/30929198241?pr=13845 -- that locally seems to also happen on `origin/main` branch itself.

**Comment by cpoerschke** [comment]
> w.r.t. the CI failure -- https://github.com/apache/lucene/actions/runs/11130252581/job/30929198241?pr=13845 -- that locally seems to also happen on `origin/main` branch itself.

merged in latest `origin/main` and the CI passes now then.

File: lucene/core/src/java/org/apache/lucene/search/similarities/DFRSimilarity.java (hunk 3 above)
Comment (rmuir) [reply lucene/core/src/java/org/apache/lucene/search/similarities/DFRSimilarity.java:101]: this isn't a DFR component. it is just an expert boolean.

File: lucene/core/src/java/org/apache/lucene/search/similarities/IBSimilarity.java (hunk 4 above)
Comment (rmuir) [reply lucene/core/src/java/org/apache/lucene/search/similarities/IBSimilarity.java:107]: same as DFR, this isn't a component

File: lucene/core/src/java/org/apache/lucene/search/similarities/LMDirichletSimilarity.java (hunk 5 above)
Comment (rmuir) [reply lucene/core/src/java/org/apache/lucene/search/similarities/LMDirichletSimilarity.java:62]: we only need one ctor here with this parameter, just add it to the "expert" ctor only.

File: lucene/core/src/java/org/apache/lucene/search/similarities/LMDirichletSimilarity.java (hunk 6 above)
Comment (rmuir) [reply lucene/core/src/java/org/apache/lucene/search/similarities/LMDirichletSimilarity.java:77]: add to expert CTOR only. no telescoping for this parameter, it is very expert.

File: lucene/core/src/java/org/apache/lucene/search/similarities/LMSimilarity.java [hunk 8]
```diff
@@ -43,11 +43,25 @@ public abstract class LMSimilarity extends SimilarityBase {
...
+  /**
+   * Creates a new instance with the default collection language model and default discountOverlaps
+   * value.
+   */
   public LMSimilarity() {
+    this(true);
+  }
+
+  /** Creates a new instance with the default collection language model. */
+  public LMSimilarity(boolean discountOverlaps) {
```
Comment (rmuir) [inline lucene/core/src/java/org/apache/lucene/search/similarities/LMSimilarity.java:64]: just needs to be added to above "expert" ctor only.

**Comment by rmuir** [comment]
I added comments, this boolean parameter is esoteric and just needs to be available on "expert ctor": not added to every possible ctor for these classes. We just need to make it possible.

File: lucene/core/src/java/org/apache/lucene/search/similarities/DFRSimilarity.java (hunk 3 above)
Comment (rmuir) [reply lucene/core/src/java/org/apache/lucene/search/similarities/DFRSimilarity.java:101]: when we use "component" in these docs, it has special meaning to the user. Let's make sure to not anything about that in these constructors (DFR is not the only one like this)

> DFR models are obtained by instantiating the three components of the framework: selecting a basic randomness model, applying the first normalisation, and normalising the term frequencies

https://en.wikipedia.org/wiki/Divergence-from-randomness_model

**Comment by cpoerschke** [comment]
> I added comments, this boolean parameter is esoteric and just needs to be available on "expert ctor": not added to every possible ctor for these classes. We just need to make it possible.

Thanks for the review! Made updates accordingly.

**Review by rmuir — APPROVED** [review]

**Pull request merged**